src/app/clusters/bindings/BindingManager.cpp

```diff
@@ -203,21 +236,33 @@ CHIP_ERROR BindingManager::NotifyBoundClusterChanged(EndpointId endpoint, Cluste
                 if (peerDevice != nullptr && peerDevice->IsConnected())
                 {
                     // We already have an active connection
-                    mBoundDeviceChangedHandler(*iter, peerDevice, context);
+                    mBoundDeviceChangedHandler(*iter, peerDevice, context->GetContext());
                 }
                 else
                 {
-                    mPendingNotificationMap.AddPendingNotification(iter.GetIndex(), context);
-                    ReturnErrorOnFailure(EstablishConnection(iter->fabricIndex, iter->nodeId));
+                    mPendingNotificationMap.AddPendingNotification(iter.GetIndex(), static_cast<void *>(context));
+                    context->IncrementConsumersNumber();
+                    error = EstablishConnection(iter->fabricIndex, iter->nodeId);
+                    SuccessOrExit(error == CHIP_NO_ERROR);
                 }
             }
             else if (iter->type == EMBER_MULTICAST_BINDING)
             {
-                mBoundDeviceChangedHandler(*iter, nullptr, context);
+                mBoundDeviceChangedHandler(*iter, nullptr, context->GetContext());
             }
         }
     }
-    return CHIP_NO_ERROR;
+
+exit:
+    context->DecrementConsumersNumber();
+
+    // Call release context handler only if any pending notification is not going to use it.
```

Comment:
So really all we are keeping track of here is whether we called `AddPendingNotification` or directly notified above, right?

Why do we need the indirection of the `BindingManagerContext` type and its reference count?

I guess an even more specific question: if applications just deleted the context in their `BoundDeviceChangedHandler`, what would go wrong?

I guess what would go wrong is if entries are removed from the map without calling the callback (e.g. if the binding is removed) we would leak... but that also leaks with this PR, because those codepaths don't call `mBoundDeviceContextReleaseHandler`.

Am I missing something?

It seems to me that we can just pass the caller-provided context to `mBoundDeviceContextReleaseHandler` and need to ensure that:
1. We call it in `NotifyBoundClusterChanged` on any success path that does not add to `mPendingNotificationMap`.
2. We either call it in `NotifyBoundClusterChanged` on failure paths or ensure that consumers know do delete their context on failure by documenting the API accordingly.
3. We call it in any codepath that removes an entry from `mPendingNotificationMap`

Reply:
> Why do we need the indirection of the BindingManagerContext type and its reference count?

My concern is that based on the `NotifyBoundClusterChanged` implementation, the if condition in for loop can be met for more than one targets. That may result in adding few notifications to the list and so in calling `HandleDeviceConnected` few times. 

> I guess an even more specific question: if applications just deleted the context in their BoundDeviceChangedHandler, what would go wrong?

If we would have more than one targets that would mean only first call will contain valid context and the others it will point to destroyed object.

Reply:
Good idea, done.